Source/WebCore/css/CSSSelector.cpp

```diff
@@ -357,6 +357,24 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
     return type;
 }
 
+std::optional<PseudoId> CSSSelector::parseStandalonePseudoElement(StringView input, const CSSSelectorParserContext& context)
+{
+    // FIXME: Tokenize input.
+    if (input.startsWith("::"_s)) {
+        auto pseudoElement = parsePseudoElement(input.substring(2), context);
+        if (!pseudoElement)
+            return std::nullopt;
+        return pseudoId(*pseudoElement);
+    }
+    if (input.startsWith(":"_s)) {
+        auto pseudoClassOrElement = parsePseudoClassAndCompatibilityElementString(input.substring(1));
+        if (!pseudoClassOrElement.compatibilityPseudoElement)
+            return std::nullopt;
+        return pseudoId(*pseudoClassOrElement.compatibilityPseudoElement);
+    }
+    return std::nullopt;
+}
+
```

Comment:
We probably have a couple of duplicated code for parsing the values of PseudoElements. Last time I checked, there are 3 places in the code. Maybe all of this should be in one place.

Reply:
The 3 you found and you are trying to fix in this PR.

* [`LocalDomWindow::getMatchedCSSRules`](https://searchfox.org/wubkat/rev/2333404a4a2e271f6ea0f797f792442db05990de/Source/WebCore/page/LocalDOMWindow.cpp#1667-1673)
* [`WebAnimationUtilities::pseudoIdFromString`](https://searchfox.org/wubkat/rev/2333404a4a2e271f6ea0f797f792442db05990de/Source/WebCore/animation/WebAnimationUtilities.cpp#340-364)
* The one I'm [working on](https://github.com/WebKit/WebKit/pull/7468/files#diff-d8731a2ff6dfd96d88c6fc6f49febb2c58020c97094acd2efc17899956913218)  from [`CSSComputedStyleDeclaration::CSSComputedStyleDeclaration`](https://searchfox.org/wubkat/rev/2333404a4a2e271f6ea0f797f792442db05990de/Source/WebCore/css/CSSComputedStyleDeclaration.cpp#50-61)

I'm happy to retire my PR. Yours is more generic. There is still the issue of what to do for the empty list of style.